test: Add regression test for memory leak (#31383)

DipayanDasgupta opened the pull request:
This PR adds a test script that definitively identifies a memory leak in the OpenVINO Python bindings, as discussed in issue #31383.
Summary of Investigation
Symptom: Initial tests showed growing RSS when loading/unloading models in a loop.
Analysis: Following maintainer feedback, a deep dive was conducted using Valgrind to get definitive proof.
Root Cause Identified: The Valgrind reports confirm a significant "still reachable" memory leak. The call stacks trace this memory back to the initialization of the Python bindings.
Proposed Fix Tested: The openvino.shutdown() function, which is intended to clean up these resources, was tested. While it may release some memory, a substantial leak of over 2.2 MB persists even after it is called.
Conclusion
This test proves there is a memory leak in the OpenVINO Python bindings where static/global resources are not being fully released by openvino.shutdown().
This PR provides a minimal, repeatable test case (simple_leak_test.py) and a README.md with instructions for using Valgrind to validate the eventual fix. This contribution will help the core team to pinpoint and resolve the underlying issue in the library's cleanup mechanism.

praasz commented:
> can it be merged?

It doesn't show is mem leak. The not RSS value maybe the reason that OS not restored the memory from process (allocator behavior)

What is the test platform for it is same behavior on Windows and Linux?

Please check also if issue can be resolved by these [hints](https://docs.openvino.ai/2025/openvino-workflow/running-inference/optimize-inference/optimizing-memory-usage.html)

If the test platform is Linux try set env variable e.g.  to `MALLOC_TRIM_THRESHOLD_=13107200`

DipayanDasgupta commented:
Hello @praasz 
Thank you for the excellent feedback. I have updated this PR to incorporate your suggestions.
I ran the test under Valgrind and have confirmed a leak of "still reachable" memor
[valgrind_report.txt](https://github.com/user-attachments/files/21492726/valgrind_report.txt)
y, which points to an issue in the Python bindings. I have updated the PR description with a full summary of these findings and have attached the complete valgrind_report.txt to this comment for your review. The branch has also been synced with the latest master.
This test should now serve as a much more robust validation tool.

praasz commented:
@DipayanDasgupta 
Hi, thank for your input the issue will be analyzed. The valgrind report is maybe leak as this still reachable may mean some static allocation, which can be still release by application.

Regarding PR:
The mergeable PR should contain fix and some reports can be provided as proof that issue is fixed (or unit test).
Right now the current PR provides analysis without fix.

DipayanDasgupta commented:
Hello @praasz,
The investigation is complete, and we have definitive proof of a memory leak.
As you suspected, the issue is with static allocations, but the key finding is that openvino.shutdown() does not completely fix it.
I have attached two Valgrind reports:
valgrind_LEAK_report.txt: Shows the memory state without calling shutdown().
valgrind_final_report.txt: Shows the memory state after calling shutdown(). A 2.2MB leak of "still reachable" memory remains.
This proves there is a bug in the shutdown cleanup process for the Python bindings. I have updated the PR with a simplified test case and a new description reflecting these findings. This should give your team everything they need to debug the C++ bindings layer.
Thank you for your guidance in narrowing this down. The PR is now ready for review as a definitive bug report and test case.
<img width="1034" height="721" alt="image" src="https://github.com/user-attachments/assets/f8ccb585-dd35-4740-abb2-1e72809b3e67" />
<img width="1211" height="709" alt="image" src="https://github.com/user-attachments/assets/3edc86c8-3a19-4e4a-944e-902db57a8688" />
